common/meta/types/src/user_auth.rs

```diff
@@ -11,17 +11,237 @@
+                PasswordType::Sha256 => Sha256Password,
+                PasswordType::DoubleSha1 => DoubleShaPassword,
+            },
+        }
+    }
+
+    pub fn get_password(&self) -> Option<Vec<u8>> {
+        match self {
+            AuthInfo::Password {
+                password: p,
+                password_type: _,
+            } => Some(p.to_vec()),
+            _ => None,
+        }
+    }
+
+    pub fn get_password_type(&self) -> Option<PasswordType> {
+        match self {
+            AuthInfo::Password {
+                password: _,
+                password_type: t,
+            } => Some(*t),
+            _ => None,
+        }
+    }
+
+    fn decode_double_sha1(
+        salt: &[u8],
+        input: &[u8],
+        user_password: &[u8],
+    ) -> Result<Vec<u8>, ErrorCode> {
+        let double_sha1 = calc_sha1(&calc_sha1(user_password)[..]);
+        // SHA1( password ) XOR SHA1( "20-bytes random data from server" <concat> SHA1( SHA1( password ) ) )
+        let mut m = sha1::Sha1::new();
+        m.update(salt);
+        m.update(&double_sha1[..]);
+
+        let result = m.digest().bytes();
+        if input.len() != result.len() {
+            return Err(ErrorCode::SHA1CheckFailed("SHA1 check failed"));
+        }
+        let mut s = Vec::with_capacity(result.len());
+        for i in 0..result.len() {
+            s.push(input[i] ^ result[i]);
+        }
+        Ok(s)
+    }
+
+    pub fn auth_mysql(&self, password_input: &[u8], salt: &[u8]) -> Result<bool, ErrorCode> {
+        match self {
+            AuthInfo::None => Ok(true),
+            AuthInfo::Password {
+                password: p,
+                password_type: t,
+            } => {
+                match t {
+                    PasswordType::PlainText => Ok(p == password_input),
+                    PasswordType::DoubleSha1 => {
+                        let password_sha1 = AuthInfo::decode_double_sha1(salt, password_input, p)?;
+                        // TODO(youngsofun): we should store double_sha1(password) later
+                        Ok(double_sha1(p) == calc_sha1(&password_sha1))
+                    }
+                    PasswordType::Sha256 => Err(ErrorCode::AuthenticateFailure(
+                        "login with sha256_password user for mysql protocol not supported yet.",
+                    )),
+                }
+            }
+        }
+    }
+
+    // for clickhouse and http basic
+    pub fn auth_password_plaintext(
+        &self,
+        password_input_plaintext: &[u8],
+    ) -> Result<bool, ErrorCode> {
+        match self {
+            AuthInfo::None => Ok(true),
+            AuthInfo::Password {
+                password: p,
+                password_type: _t,
+                // TODO(youngsofun): we should store password hash later
+            } => Ok(p == password_input_plaintext),
+        }
+    }
+}
+
+#[derive(Clone, serde::Serialize, serde::Deserialize, Debug, PartialEq)]
+pub struct AuthInfoRaw {
```

Comment:
`AuthInfoRaw` confused here with `AuthInfo`, maybe we need a better name :thinking:

Reply:
AuthInfoRaw means it is directly mapped to SQL.
usage of it depends on the AuthType.
I choose to not make further abstraction to make it clear to the implementer of AuthType.
what about AuthInfoArgs?

Reply:
`*_raw` -> `*_args` also need changed